File: src/main/java/com/skysoft/mixin/MouseHandlerMixin.java
package com.skysoft.mixin;

import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
import com.mojang.blaze3d.platform.Window;
import com.skysoft.features.bazaar.BazaarTracker;
import com.skysoft.gui.scale.CursorController;
import com.skysoft.gui.scale.GuiScaleController;
import com.skysoft.gui.scale.InventoryCursorMemory;
import com.skysoft.utils.MinecraftClient;
import com.skysoft.utils.input.InputHandlingResult;
import net.minecraft.client.MouseHandler;
import net.minecraft.client.input.MouseButtonInfo;
import org.objectweb.asm.Opcodes;
import org.lwjgl.glfw.GLFW;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@Mixin(MouseHandler.class)
public class MouseHandlerMixin implements CursorController {
    @Shadow
    private double xpos;

    @Shadow
    private double ypos;

    @Override
    public void skysoft$moveCursor(double cursorX, double cursorY) {
        xpos = cursorX;
        ypos = cursorY;
    }

    @Inject(
        method = "grabMouse",
        at = @At(
            value = "FIELD",
            target = "Lnet/minecraft/client/MouseHandler;mouseGrabbed:Z",
            opcode = Opcodes.PUTFIELD
        )
    )
    private void skysoft$saveCursorBeforeGrab(CallbackInfo ci) {
        InventoryCursorMemory.beginMouseGrab(xpos, ypos);
    }

    @Inject(
        method = "grabMouse",
        at = @At(
            value = "INVOKE",
            target = "Lcom/mojang/blaze3d/platform/InputConstants;grabOrReleaseMouse(Lcom/mojang/blaze3d/platform/Window;IDD)V"
        )
    )
    private void skysoft$saveCursorAfterGrab(CallbackInfo ci) {
        InventoryCursorMemory.finishMouseGrab(xpos, ypos);
    }

    @Inject(method = "onButton", at = @At("HEAD"), cancellable = true)
    private void skysoft$clickBazaarTrackerControl(long window, MouseButtonInfo buttonInfo, int action, CallbackInfo ci) {
        if (
            action == GLFW.GLFW_PRESS
                && BazaarTracker.handleMouseButtonPress(buttonInfo.button()) == InputHandlingResult.CONSUMED
        ) {
            ci.cancel();
        }
    }

    @WrapOperation(
        method = "releaseMouse",
        at = @At(
            value = "INVOKE",
            target = "Lcom/mojang/blaze3d/platform/InputConstants;grabOrReleaseMouse(Lcom/mojang/blaze3d/platform/Window;IDD)V"
        )
    )
    private void skysoft$restoreCursorOnRelease(
        Window window,
        int cursorMode,
        double cursorX,
        double cursorY,
        Operation<Void> original
    ) {
        InventoryCursorMemory.CursorPoint restored = InventoryCursorMemory.cursorForRelease(cursorX, cursorY);
        if (restored != null) {
            cursorX = restored.x();
            cursorY = restored.y();
            xpos = cursorX;
            ypos = cursorY;
        }

        original.call(window, cursorMode, cursorX, cursorY);
    }

    @Inject(method = "getScaledXPos(Lcom/mojang/blaze3d/platform/Window;D)D", at = @At("HEAD"), cancellable = true)
    private static void skysoft$getInventoryScaledX(Window window, double xPosition, CallbackInfoReturnable<Double> cir) {
        if (!GuiScaleController.scalesInventory(MinecraftClient.screen()) ||
            GuiScaleController.overlaysUseNormalCoordinates()) {
            return;
        }

        try (GuiScaleController.WindowScaleOverride ignored =
                 GuiScaleController.useInventoryScale(MinecraftClient.screen(), window)) {
            cir.setReturnValue(xPosition * window.getGuiScaledWidth() / (double) window.getScreenWidth());
        }
    }

    @Inject(method = "getScaledYPos(Lcom/mojang/blaze3d/platform/Window;D)D", at = @At("HEAD"), cancellable = true)
    private static void skysoft$getInventoryScaledY(Window window, double yPosition, CallbackInfoReturnable<Double> cir) {
        if (!GuiScaleController.scalesInventory(MinecraftClient.screen()) ||
            GuiScaleController.overlaysUseNormalCoordinates()) {
            return;
        }

        try (GuiScaleController.WindowScaleOverride ignored =
                 GuiScaleController.useInventoryScale(MinecraftClient.screen(), window)) {
            cir.setReturnValue(yPosition * window.getGuiScaledHeight() / (double) window.getScreenHeight());
        }
    }
}

File: src/main/kotlin/com/skysoft/mixin/MouseHandlerMixin.kt
package com.skysoft.mixin

import com.llamalad7.mixinextras.injector.wrapoperation.Operation
import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation
import com.mojang.blaze3d.platform.Window
import com.skysoft.features.bazaar.BazaarTracker
import com.skysoft.features.inventory.StorageOverlayController
import com.skysoft.gui.scale.CursorController
import com.skysoft.gui.scale.GuiScaleController
import com.skysoft.gui.scale.InventoryCursorMemory
import com.skysoft.gui.tooltip.TooltipViewport
import com.skysoft.utils.MinecraftClient
import com.skysoft.utils.input.InputHandlingResult
import net.minecraft.client.MouseHandler
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen
import net.minecraft.client.input.MouseButtonInfo
import org.objectweb.asm.Opcodes
import org.lwjgl.glfw.GLFW
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

@Mixin(MouseHandler::class)
open class MouseHandlerMixin : CursorController {
    @field:Shadow
    private var xpos = 0.0

    @field:Shadow
    private var ypos = 0.0

    override fun skysoftMoveCursor(cursorX: Double, cursorY: Double) {
        xpos = cursorX
        ypos = cursorY
    }

    @Inject(
        method = ["grabMouse"],
        at = [
            At(
                value = "FIELD",
                target = "Lnet/minecraft/client/MouseHandler;mouseGrabbed:Z",
                opcode = Opcodes.PUTFIELD,
            ),
        ],
    )
    protected fun skysoftSaveCursorBeforeGrab(ci: CallbackInfo) {
        InventoryCursorMemory.beginMouseGrab(xpos, ypos)
    }

    @Inject(
        method = ["grabMouse"],
        at = [
            At(
                value = "INVOKE",
                target = "Lcom/mojang/blaze3d/platform/InputConstants;grabOrReleaseMouse(" +
                    "Lcom/mojang/blaze3d/platform/Window;IDD)V",
            ),
        ],
    )
    protected fun skysoftSaveCursorAfterGrab(ci: CallbackInfo) {
        InventoryCursorMemory.finishMouseGrab(xpos, ypos)
    }

    @Inject(method = ["onButton"], at = [At("HEAD")], cancellable = true)
    protected fun skysoftClickBazaarTrackerControl(
        window: Long,
        buttonInfo: MouseButtonInfo,
        action: Int,
        ci: CallbackInfo,
    ) {
        if (
            action == GLFW.GLFW_PRESS &&
            BazaarTracker.handleMouseButtonPress(buttonInfo.button()) == InputHandlingResult.CONSUMED
        ) {
            ci.cancel()
        }
    }

    @WrapOperation(
        method = ["onScroll"],
        at = [
            At(
                value = "INVOKE",
                target = "Lnet/minecraft/client/gui/screens/Screen;mouseScrolled(DDDD)Z",
            ),
        ],
    )
    protected fun doesSkysoftHandleTooltipScroll(
        screen: Screen,
        mouseX: Double,
        mouseY: Double,
        horizontalAmount: Double,
        verticalAmount: Double,
        original: Operation<Boolean>,
    ): Boolean {
        val container = screen as? AbstractContainerScreen<*>
        val overStorageScrollPanel = container != null &&
            StorageOverlayController.shouldPreferMouseScroll(container, mouseX, mouseY, verticalAmount)
        if (overStorageScrollPanel) {
            if (
                TooltipViewport.isStorageOverlayScrollKeyDown() &&
                TooltipViewport.didHandleStorageMouseScroll(horizontalAmount, verticalAmount)
            ) {
                return true
            }
            return original.call(screen, mouseX, mouseY, horizontalAmount, verticalAmount)
        }
        if (TooltipViewport.didHandleMouseScroll(horizontalAmount, verticalAmount)) return true
        return original.call(screen, mouseX, mouseY, horizontalAmount, verticalAmount)
    }

    @WrapOperation(
        method = ["releaseMouse"],
        at = [
            At(
                value = "INVOKE",
                target = "Lcom/mojang/blaze3d/platform/InputConstants;grabOrReleaseMouse(" +
                    "Lcom/mojang/blaze3d/platform/Window;IDD)V",
            ),
        ],
    )
    protected fun skysoftRestoreCursorOnRelease(
        window: Window,
        cursorMode: Int,
        initialCursorX: Double,
        initialCursorY: Double,
        original: Operation<Void>,
    ) {
        var cursorX = initialCursorX
        var cursorY = initialCursorY
        InventoryCursorMemory.cursorForRelease(cursorX, cursorY)?.let { restored ->
            cursorX = restored.x
            cursorY = restored.y
            xpos = cursorX
            ypos = cursorY
        }
        original.call(window, cursorMode, cursorX, cursorY)
    }

    private companion object {
        @JvmStatic
        @Inject(
            method = ["getScaledXPos(Lcom/mojang/blaze3d/platform/Window;D)D"],
            at = [At("HEAD")],
            cancellable = true,
        )
        private fun skysoftGetInventoryScaledX(
            window: Window,
            xPosition: Double,
            cir: CallbackInfoReturnable<Double>,
        ) {
            val screen = MinecraftClient.screen()
            if (
                !GuiScaleController.usesSeparateInventoryScale(screen) ||
                GuiScaleController.areOverlaysUsingNormalCoordinates()
            ) {
                return
            }

            GuiScaleController.useInventoryScale(screen, window).use {
                cir.setReturnValue(xPosition * window.guiScaledWidth / window.screenWidth.toDouble())
            }
        }

        @JvmStatic
        @Inject(
            method = ["getScaledYPos(Lcom/mojang/blaze3d/platform/Window;D)D"],
            at = [At("HEAD")],
            cancellable = true,
        )
        private fun skysoftGetInventoryScaledY(
            window: Window,
            yPosition: Double,
            cir: CallbackInfoReturnable<Double>,
        ) {
            val screen = MinecraftClient.screen()
            if (
                !GuiScaleController.usesSeparateInventoryScale(screen) ||
                GuiScaleController.areOverlaysUsingNormalCoordinates()
            ) {
                return
            }

            GuiScaleController.useInventoryScale(screen, window).use {
                cir.setReturnValue(yPosition * window.guiScaledHeight / window.screenHeight.toDouble())
            }
        }
    }
}
